### ISSUE TITLE
Discuss: Remove all Bump Mapping and Parallax Occlusion code

### ISSUE THREAD

**Issue by lhofhansl** [minetest/minetest#10479]
As I look through the code - prompted by the inclusion of the bump-mapping test nodes into devtest - I'm coming more and more to the conclusion that the code should just be removed until we find someone who knows (and has time) to implement this for real.

Issues I've seen:
- normal maps are not translated from tangent space (i.e. are not adjusted for face's direction). There is an attempt it seems with calculating a tsEyeVec and tsLightVec in the vertex shader.
- wrong vectors used in the fragment shader
- diffuse light calculation is doubly wrong (uses world-eye vector, and should use the tangent-space light vector in the first place, diffuse light is independent of the position of the viewer)

Then there are other issues such as hardcoded lighting of the faces in the engine, which is at odds with lighting in the shader, since now you'd have to adjust for that.

I played around with this for a few hours, but it looks like there are other problems. Normals missing (or wrong), etc.

In the end, as much as it pains me (I'm a fan of better graphics), I think we might want to simply remove all bump mapping and related code, keep it simple. That way we also simplify the shader code and maybe provide a cleaner slate for trying again. 

Thoughts?

**Comment by LoneWolfHT** [comment]
I don't use either of them. Looks ugly IMO

**Comment by kilbith** [comment]
TL;DR: RBA shaders is half-assed work and must be redone from scratch (example: tonemapping is a postprocessing thing normally)

**Comment by lhofhansl** [comment]
@kilbith I agree with the state of the shaders.

I'm trying to undo at least the parts that do not work, so we have less to redo.
With bump mapping et al code removed there really isn't much left, especially in the fragment shader.

Re: RBA (RIP)... I will say that he just didn't get to finish them. Someone (forgot who) showed me some video of his uncommitted work a while ago and that looked really impressive.

Related: Wasn't there some postprocessing PR? What happened to that one.

**Comment by kilbith** [comment]
> Related: Wasn't there some postprocessing PR? What happened to that one.

https://github.com/minetest/minetest/pull/3623
https://github.com/minetest/minetest/pull/5913
https://github.com/minetest/minetest/pull/5932
https://github.com/minetest/minetest/pull/9895

Though none of them properly utilize the GPU framebuffers that I used in https://github.com/minetest/minetest/pull/6839.

**Comment by paramat** [comment]
:+1:

Bumpmapping looks so bad i think it is better to have none until it is done well than leave it in months or years waiting to be fixed and giving users a bad impression.
As you can see in #9241 i discovered very broken bumpmapping behaviour.
Few users actually use normalmap textures, most use the hideous 'generate normalmaps' which has already been removed, so most users will be losing bumpmapping anyway.

Parallax occlusion seemed to work and looked more acceptable to me. But i do not know the code as well as you do. The code may be closely linked to the bumpmapping code. It might help to make a fresh start with that too. So, i would support removing that too if you judge that to be best.

**Comment by lhofhansl** [comment]
Occlusion not exclusion. :)

Ok... I'll file a PR to remove all the code, including the UI and the various options passed to the shaders.

**Comment by lhofhansl** [comment]
@HybridDog Any opinion?

**Comment by lhofhansl** [comment]
I was under the impression that parallax occlusion does not work without bump mapping.
Lemme look and try to disentangle the two.

Can parallax occlusion work without correct normals?

**Comment by paramat** [comment]
> I am still in favor removing both and starting again

Me too.

I did some testing of parallax occlusion. When more than 1 node away the node top face seems to appear ok, but note that the other node faces are inconsistent and much flatter, and partial pixels from the opposite edge are present:

![screenshot_20201013_174619](https://user-images.githubusercontent.com/3686677/95898581-28221380-0d87-11eb-9526-3e0c51d86f11.png)

Looking at the node top face, as your view angle decreases the height of the raised pixels decreases relative to their widths:

![screenshot_20201013_174642](https://user-images.githubusercontent.com/3686677/95900707-358ccd00-0d8a-11eb-8f23-03a704ebbdbe.png)

So the appearence is messy and has weird behaviour.

Up close the raised pixels 'lean' in odd ways, although this is much more subtle at typical distances, so not much of an issue. It is also not surprising since this is 'faked' 3D structure:

![screenshot_20201013_174728](https://user-images.githubusercontent.com/3686677/95898761-73d4bd00-0d87-11eb-80fe-57496a104195.png)

**Comment by paramat** [comment]
> I think this is supposed to happen unless the mod explicitly disables it for the tiles with tileable_vertical = false and/or tileable_horizontal = false.

Yes. The wrapping pixels are needed for an area of nodes, but look wrong for one node. The 'tileable' flags force one behaviour or another, regardless of whether a node is alone or in an area of nodes, so they are not a complete solution.
However, this is an unsolvable problem, not bad shader coding.
> This may be intentional, too.

I doubt it, sloppiness is far more likely.
It looks really bad. Even if it was intentional it was a poor decision, the wrapping pixels are less of a visual problem.

**Comment by lhofhansl** [comment]
#10487 is merged.

**Comment by lhofhansl** [comment]
@HybridDog Happy to work together to add graphics features back in.
